lib/storage/data/external/GCP/GcpService.js

```diff
@@ -1,141 +1,889 @@
+ *   - MPU (multipart upload) is emulated using GCP's compose/copy APIs.
+ *   - All S3-compatible methods are available, with overrides for GCP quirks.
+ */
+class GcpClient extends S3Client {
+    /**
+     * @param {object} config - Configuration object (see above)
+     */
+    constructor(config) {
+        super(config.s3Params);
+        this._config = config;
+        this._maxConcurrent = 10;
 
-Object.assign(GCP.prototype, GcpServiceSetup, {
-    _maxConcurrent: 5,
+        this.middlewareStack.remove('httpSigningMiddleware');
 
-    // Implemented APIs
-    // Bucket API
-    getBucket(params, callback) {
-        return this.listObjects(params, callback);
-    },
+        // Normalize headers before signing to avoid non-string values
+        this.middlewareStack.add(
+            createNormalizeHeadersMiddleware(),
+            {
+                name: 'normalizeHeadersMiddleware',
+                step: 'build',
+                priority: 'high',
+            }
+        );
 
-    // Object APIs
-    upload(params, callback) {
-        try {
-            const uploader = new GcpManagedUpload(this, params);
-            return uploader.send(callback);
-        } catch (err) {
-            return callback(err);
-        }
-    },
+        // Add DNS compatibility middleware AFTER bucket endpoint resolution
+        // This runs after the SDK decides virtual-hosted vs path-style
+        // and forces path-style for DNS-incompatible buckets
+        this.middlewareStack.add(
+            gcpDnsCompatibilityMiddleware(config),
+            {
+                name: 'gcpDnsCompatibility',
+                step: 'serialize',
+                priority: 'high'
+            }
+        );
+
+        // Add GCP signing middleware
+        this.middlewareStack.add(
+            gcpSigningMiddleware(config),
+            {
+                name: 'gcpSigningMiddleware',
+                step: 'finalizeRequest',
+                priority: 'high'
+            }
+        );
+    }
 
-    putObjectCopy(params, callback) {
-        return this.copyObject(params, callback);
-    },
+    /**
+     * Upload an object to GCP (with optional tagging as metadata).
+     * @param {object} params - S3 putObject params
+     * @param {function} callback
+     */
+    putObject(params, callback) {
+        const command = new PutObjectCommand(params);
+        
+        // Capture GCP-specific headers
+        let objectGeneration;
+        command.middlewareStack.add(
+            next => async args => {
+                const result = await next(args);
+                const generationHeader = getHeaderCaseInsensitive(
+                    result.response?.headers,
+                    'x-goog-generation'
+                );
+                if (generationHeader !== undefined) {
+                    objectGeneration = generationHeader;
+                }
+                return result;
+            },
+            {
```

Comment:
this is duplicated in every call, if we can't move it to the client's Middlewarestack we must anyway make it into a reusable function